src/run-state-uploader.ts

```diff
@@ -0,0 +1,57 @@
+import { RunState } from './types/run';
+import { UploadInfo } from './types/internal';
+
+export const RUN_STATE_UPLOAD_PERIOD = 10000;
+
+export interface Uploader {
+    getUploadInfo (uploadEntityId: string): Promise<UploadInfo>;
+    upload (uploadUrl: string, uploadEntity: Buffer): Promise<void>;
+}
+
+export interface RunStateProvider {
+    getRunState(): Partial<RunState>;
+}
+
+function skipEmptyValues (key, value) {
+    return value instanceof RegExp ? value.toString() : value;
+}
+
+export class RunStateUploader {
+    private _runStateProvider: RunStateProvider;
+    private _uploader: Uploader;
+    private _intervalId: NodeJS.Timeout;
+    private _uploadInfo: UploadInfo;
+    private _runStateUploadPeriod: number;
+
+    async _uploadRunState () {
+        const runState = this._runStateProvider.getRunState();
+
+        const buffer = Buffer.from(JSON.stringify(runState, skipEmptyValues));
+
+        this._uploader.upload(this._uploadInfo.uploadUrl, buffer);
+    }
+
+    constructor (uploader: Uploader, runStateProvider: RunStateProvider, runStateUploadPeriod = RUN_STATE_UPLOAD_PERIOD) {
+        this._uploader = uploader;
+        this._runStateProvider = runStateProvider;
+        this._runStateUploadPeriod = runStateUploadPeriod;
+
+        this._uploadRunState = this._uploadRunState.bind(this);
+    }
+
+    async start (runId: string) {
+        this._uploadInfo = await this._uploader.getUploadInfo(runId);
```

Comment:
This class instance works with the same run state provider for its entire lifetime. `RunStateBuilder`, which is currently the only run state provider we have, is bound to a `runId` in its constructor, so it also cannot change during an instance lifetime.

So, what is the point of passing different `runIds` to the `start()` method? Maybe we can take it from the run state provider or specify in a constructor?

Reply:
Don't sure about API for this class yet. i think we should clarify it when we combine all the parts togther